staging/src/k8s.io/client-go/tools/dnscache/dialer.go

```diff
@@ -0,0 +1,283 @@
+	DNSQuery           int64
+	SuccessfulDNSQuery int64
+}
+
+type dialer struct {
+	dialer            Dialer
+	resolver          Resolver
+	minCacheDuration  time.Duration
+	maxCacheDuration  time.Duration
+	forceRefreshTimes int64
+	rand              *rand.Rand
+
+	lock  sync.RWMutex
+	cache map[string]*cacheItem
+
+	chanLock        sync.Mutex
+	resolveChannels map[string]chan error
+
+	stats DialerStats
+}
+
+// NewDialer creates a dialer with dns cache.
+func NewDialer(config *DialerConfig) (Dialer, error) {
+	d := &dialer{
+		dialer:            config.Dialer,
+		resolver:          config.Resolver,
+		minCacheDuration:  config.MinCacheDuration,
+		maxCacheDuration:  config.MaxCacheDuration,
+		forceRefreshTimes: config.ForceRefreshTimes,
+		rand:              rand.New(rand.NewSource(time.Now().UnixNano())),
+		cache:             map[string]*cacheItem{},
+		resolveChannels:   map[string]chan error{},
+	}
+	if d.dialer == nil {
+		d.dialer = &net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}
+	}
+	if d.resolver == nil {
+		d.resolver = &net.Resolver{}
+	}
+	if d.minCacheDuration == 0 {
+		d.minCacheDuration = time.Second * 30
+	}
+	if d.maxCacheDuration == 0 {
+		d.maxCacheDuration = time.Minute
+	}
+	if d.forceRefreshTimes == 0 {
+		d.forceRefreshTimes = 10
+	}
+	if d.forceRefreshTimes < 0 {
+		d.forceRefreshTimes = 0
+	}
+	if d.minCacheDuration > d.maxCacheDuration {
+		return nil, fmt.Errorf("min cache duration(%s) should less than or equal to max cache duration(%s)", d.minCacheDuration.String(), d.maxCacheDuration.String())
+	}
+	return d, nil
+}
+
+func (d *dialer) Stats() DialerStats {
+	return d.stats
+}
+
+func (d *dialer) Dial(network, address string) (net.Conn, error) {
+	return d.DialContext(context.Background(), network, address)
+}
+
+func (d *dialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	atomic.AddInt64(&d.stats.TotalConn, 1)
+	if d.forceRefreshTimes > 0 && (network == "tcp" || network == "tcp4") && address != "" {
+		host, port, err := net.SplitHostPort(address)
+		if err != nil {
+			return nil, err
+		}
+		if host != "" {
+			ip, err := d.resolveHost(ctx, host)
+			if err == nil {
+				atomic.AddInt64(&d.stats.CacheHit, 1)
+				// Rewrite address to ip:port.
+				address = net.JoinHostPort(ip.String(), port)
+			} else {
+				atomic.AddInt64(&d.stats.CacheMiss, 1)
+				return nil, err
+			}
+		}
+	}
+	return d.dialer.DialContext(ctx, network, address)
+}
+
+func (d *dialer) resolveHost(ctx context.Context, host string) (net.IPAddr, error) {
+	ip, ok := d.getIPFromCache(ctx, host)
+	if ok {
+		return ip, nil
+	}
+
+	d.chanLock.Lock()
+	ch := d.resolveChannels[host]
+	if ch == nil {
+		// Recheck the cache.
+		ip, ok := d.getIPFromCache(ctx, host)
+		if ok {
+			return ip, nil
+		}
+		ch = make(chan error, 1)
+		d.resolveChannels[host] = ch
+		// There is no resolving process for the host. Create a new goroutine to lookup dns.
+		go func() {
```

Comment:
I think sync's Cond type is a more natural way of solving problems like this than this channel/goroutine combo. This whole function looks like it would be vastly more testable if the pieces were broken apart into multiple functions.

Reply:
`sync.Cond` and `sync.WaitGroup` don't support `select`.

Reply:
Yeah I see now you're trying to integrate with ctx.Done(), which makes it pretty hard. Even with goroutines & channels, though, I think this problem can usually be solved without chaining an error through a channel to a bunch of waiting go routines one at a time. I am too tired to suggest a model right now lest I make my comments even less useful than they already have been :/